feat(visibility): introduce partial visibility state

mediavrog opened the pull request:
fixes #838

TODO
- add tests
  - [x] threshold 0 (no events) a30c82f2986680ad24998d31f0b6dd4805d5cad7 enough? should build a test matrix
  - [x] threshold 50 (fire visible/invisible events) b7a7ba03ddc074c50d55f7410449fee09b975930
  - [ ] threshold 99 but element bigger than viewport -> for now should not trigger events

mediavrog committed:
initial commit

eboudrant commented on epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java:
```diff
@@ -100,6 +102,15 @@ public void setOnChangedEnabled(boolean enabled) {
+   *
+   * @param threshold Percentage of visible area of an element. 0..100.
+   */
+  public void setPartialImpressionThreshold(int threshold) {
```
We could add `@IntRange(from = 0, to = 100)` to the param and I wonder if we should even name it `thresholdPercentage`

mediavrog replied:
`thresholdPercentage` sounds good

mediavrog replied:
832f83d3fbb076502d94184d6361b722d2966bab

eboudrant commented on epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java:
```diff
@@ -153,6 +164,13 @@ private boolean isInFocusVisible() {
+    final int totalArea = height * width;
+    final int visibleArea = visibleHeight * visibleWidth;
+
+    return (visibleArea / (float)totalArea) * 100 >= threshold;
```
Looks like we should not send event if `threshold` is 0 (default value) ?

mediavrog replied:
good point

mediavrog replied:
832f83d3fbb076502d94184d6361b722d2966bab

eboudrant reviewed (commented):
Thanks for adding this !

In addition to that unit test will be nice and if you want to go further you can also update the `OnVisibilityEventDrawable.kt` in the `kotlinsample/` app.

eboudrant commented on epoxy-adapter/src/main/java/com/airbnb/epoxy/VisibilityState.java:
```diff
@@ -45,4 +45,12 @@
+   * greater than a specified threshold. The threshold can be set in
+   * {@link EpoxyVisibilityTracker#setPartialImpressionThreshold(int)}.
+   */
+  public static final int PARTIAL_IMPRESSION_VISIBLE = 5;
```
If you look at the `FOCUSED_VISIBLE` it protect against component that are bigger than the view port. 

```
  /**
   * Event triggered when a Component enters the Focused Range. This happens when either the
   * Component occupies at least half of the viewport or, if the Component is smaller than half of
   * the viewport, when the it is fully visible.
   */
  public static final int FOCUSED_VISIBLE = 2;
```

I was wondering if you want to implements the same protection?

The problem is, if one of the component is bigger than view port (or if threshold % bigger that view port) you will never receive a `PARTIAL_IMPRESSION_VISIBLE` event.

mediavrog replied:
I'll have a look into `FOCUSED_VISIBLE` and it's protection

mediavrog replied:
> The problem is, if one of the component is bigger than view port (or if threshold % bigger that view port) you will never receive a PARTIAL_IMPRESSION_VISIBLE event.

What do you think users would expect? I see some possible ways of dealing with this:

1) Do actually define it as accurate behavior that no event is fired if the threshold cannot be met. The events would more likely to be fired if the threshold is small. Event firing cannot be guaranteed though as you mentioned.
2) If the element is bigger than the viewport, change the frame of reference for the threshold to relate to the viewport instead of the element (so with a 50% threshold the event would fire once the element covers 50% of the viewport). The question here is if this two-fold behavior could be confusing to users of the lib or just plain undesired.
3) Hybrid approach: For as long as the threshold can be met, fire it. In the case the threshold cannot be reached, fired it if the view covers the whole viewport. 

@eboudrant From your experience with epoxy in use and feedback so far, which of these behaviors (or which other suggestion) would be most appropriate?

mediavrog replied:
Another comment: I looked at the code for `FULL_IMPRESSION_VISIBLE`, which also does not handle the case if an element is bigger than the viewport. I tend towards keeping this same behavior for the `PARTIAL_IMPRESSION_VISIBLE` as well as it would be consistent and predictable. (referring the list above it would be behavior `1.`)

In the case with `FOCUSED_VISIBLE`, the definition of what is `focused` gives some leeway for interpretation, so I think in that case having a fluid focused range is appropriate.

mediavrog committed:
rename threshold to thresholdPercentage

mediavrog committed:
pass partial visibility threshold to child recycler views

mediavrog committed:
update kotlin sample to handle partial state; sample uses 75 percent

mediavrog committed:
introduce PARTIAL_IMPRESSION_INVISIBLE

mediavrog committed:
be more specific in doc about when threshold is met for PARTIAL_IMPRESSION_VISIBLE

mediavrog committed:
remove debug statement

mediavrog committed:
kotlinsample: move partial state circle right next to visible

mediavrog committed:
basic tests for partial visibility

mediavrog committed:
test without partial viz

mediavrog commented:
@eboudrant  Could you also please have a look at the added tests?
a30c82f2986680ad24998d31f0b6dd4805d5cad7

Since setting the threshold will add a new dimension to the test matrix, I was not sure if I should adjust the whole testsuite to cope for this. Basically all of
- testDataAttachedToRecyclerView
- testScrollToPosition
- testMoveDataDown
- testInsertData
- testScrollBy
- testDeleteData
- testMoveDataUp

could be tested once with a threshold set, and once without. 
Please let me know if I should cope for this

mediavrog commented:
Also, in https://github.com/smartnews/epoxy/pull/1/commits/a30c82f2986680ad24998d31f0b6dd4805d5cad7 I added some tolerance to the `AssertHelper`:

1) `TOLERANCE_PIXELS` (value=1) was currently not being tolerant, as it was compared with `<` instead of `<=`
2) for `visibleHeight` and `visibleWidth` I added a delta to cover the `49%` case to test for an element that got `VISIBLE`, but not `PARTIAL_IMPRESSION_VISIBLE`

mediavrog committed:
remove unused reset of partial viz threshold

mediavrog committed:
format code

mediavrog committed:
disable partial impression via null and handle 0 like VISIBLE
